docs/contributing/writing-instrumentation-module.md

```diff
@@ -360,3 +360,203 @@ For example:
+
+With indy advices however, the advice method signature can't introduce types that are only
+visible in the advice classes.
+In practice, two sets of types can be used in the advice method signature:
+- types from the JDK in the bootstrap classloader like `Object`
+- types already visible from the instrumented code
+
+For example, when trying to instrument a method that has `HttpServletRequest` as argument:
+- the `HttpServletRequest` type can be used as it's part of the instrumented code
+- the `HttpServletRequestScope` is part of instrumentation and has to be of type `Object` for both the enter return value and the exit argument.
+
+```java
+@Advice.OnMethodEnter(suppress = Throwable.class, inlined = false)
+public static Object onEnter(@Advice.Argument(1) HttpServletRequest request) {
+  // custom type from the instrumentation module, has to be returned as Object
+  return new HttpServletRequestScope(request);
+}
+
+@Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inlined = false)
+public static void onExit(@Advice.Argument(1) Object request,
+                          @Advice.Enter Object enterScope) {
+  // enter scope provided as Object so we have to check type and cast
+  if (!(enterScope instanceof HttpServletRequestScope)) {
+    return;
+  }
+  HttpServletRequestScope requestScope = (HttpServletRequestScope)enterScope;
+}
+```
+
+### advice local variables
+
+With inlined advices, declaring an advice method argument with `@Advice.Local` allows defining
+a variable that is local to the advice execution for communication between the enter and exit advices.
+
+When advices are not inlined, usage of `@Advice.Local` is not possible. It is however possible to
+return a value from the enter advice and get the value in the exit advice with a parameter annotated
+with `@Advice.Enter`, for example:
+
+```java
+@Advice.OnMethodEnter(suppress = Throwable.class, inlined = false)
+public static Object onEnter(@Advice.Argument(1) Object request) {
+  return "enterValue";
+}
+
+@Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inlined = false)
+public static void onExit(@Advice.Argument(1) Object request,
+                          @Advice.Enter Object enterValue) {
+  // do something with enterValue
+}
+```
+
+### modifying method arguments
+
+With inlined advices, using the `@Advice.Argument` annotation on method parameter with `readOnly = false`
+allows modifying instrumented method arguments.
+
+When using non-inlined advices, reading the argument values is still done with `@Advice.Argument`
+annotated parameters, however modifying the values is done through the advice method return value
+and `@Advice.AssignReturned.ToArguments` annotation:
+
+The advice method return value must be `Object` unless it's a type that is provided by the JDK.
+Also, the `typing = Assigner.Typing.DYNAMIC` is needed to instruct ByteBuddy to properly cast the
```

Comment:
In #11884 I'll propose to force dynamic assignment without having to repeat it on each advice. I'll update the docs if this PR gets merged first or propose some changes here stating this fact just for information purposes if #11884 makes it in first.

Reply:
+1 on making it the dynamic assignment the default here as it makes it simpler for instrumentation authors and debugging it can be challenging.